### PR TITLE
Add a signature verification function 

### PR DESCRIPTION
**Pull request by gingerhot**
Some changes I made:
* Add a signature verification function
* Add a simple testing to session package
* Reformat some underscore variables to camel format 
* Update the README file

### CODE REVIEW

File: session/session.go [hunk 1]
```diff
@@ -46,19 +65,26 @@ func decryptCookie(cookie []byte, secret []byte) (dd []byte, err error) {
...
-	cookie, err := url.QueryUnescape(signed_cookie)
+func DecryptSignedCookie(signedCookie, secretKeyBase, salt, signSalt string) (session []byte, err error) {
+	cookie, err := url.QueryUnescape(signedCookie)
 	if err != nil {
 		return
 	}
 
 	vectors := strings.SplitN(cookie, "--", 2)
+	if vectors[0] == "" || vectors[1] == "" {
+		return nil, errors.New("Invalid cookie!")
```
Comment (andrewslotin) [inline session/session.go]: Let's use lower case and prepend the error string with package name to follow the [proposed error format](https://blog.golang.org/error-handling-and-go). Also I'm not sure if we should use `!` here.

File: session/session.go [hunk 2]
```diff
@@ -46,19 +65,26 @@ func decryptCookie(cookie []byte, secret []byte) (dd []byte, err error) {
...
+	}
+	if verified := verifySign(vectors[0], vectors[1], secretKeyBase, signSalt); !verified {
+		return nil, errors.New("Not verified signature!")
```
Comment (andrewslotin) [inline session/session.go]: Same as above. Also I'd probably consider returning a sentinel error here, so the caller could easily distinguish between the input error and invalid signature cases.

File: session/session_test.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	}
+	// a fault signature case
+	faultSignSalt := "fault signature salt"
+	if verified := verifySign(vectors[0], vectors[1], secretKeyBase, faultSignSalt); verified {
+		t.Error("The function verifySign() test failure!")
+	}
+}
+
+func TestDecryptSignedCookie(t *testing.T) {
+	cookieData, _ := DecryptSignedCookie(signedCookie, secretKeyBase, salt, signSalt)
```
Comment (andrewslotin) [inline session/session_test.go]: You might also want to assert on `err == nil` here to get a meaningful test failure once/if the signature verification fails.

File: session/session.go [hunk 4]
```diff
@@ -3,16 +3,35 @@ package session
...
 
 	"golang.org/x/crypto/pbkdf2"
 )
 
 func generateSecret(base, salt string) []byte {
-	return pbkdf2.Key([]byte(base), []byte(salt), key_iter_num, key_size, sha1.New)
+	return pbkdf2.Key([]byte(base), []byte(salt), keyIterNum, keySize, sha1.New)
+}
+
+// I stolen the snippet from: https://gist.github.com/doitian/2a89dc9e4372e55335c9111f576b47bf
```
Comment (andrewslotin) [inline session/session.go]: Probably it would be better to rephrase it to be something like this:
```go
// The origin of this snippet can be found at https://gist.github.com/doitian/2a89dc9e4372e55335c9111f576b47bf
```
😉

File: session/session.go [hunk 5]
```diff
@@ -3,16 +3,35 @@ package session
...
+func verifySign(encryptedData, sign, base, signSalt string) bool {
+	signKey := generateSecret(base, signSalt)
+	signHmac := hmac.New(sha1.New, signKey)
+	signHmac.Write([]byte(encryptedData))
+	verifySign := signHmac.Sum(nil)
+	signDecoded, err := hex.DecodeString(sign)
+	if err != nil {
+		return false
+	}
+	if hmac.Equal(verifySign, signDecoded) {
```
Comment (andrewslotin) [inline session/session.go]: I guess `return hmac.Equal(verifySign, signDecoded)` would be enough here.

File: session/session.go [hunk 6]
```diff
@@ -3,16 +3,35 @@ package session
...
+
+// I stolen the snippet from: https://gist.github.com/doitian/2a89dc9e4372e55335c9111f576b47bf
+func verifySign(encryptedData, sign, base, signSalt string) bool {
+	signKey := generateSecret(base, signSalt)
+	signHmac := hmac.New(sha1.New, signKey)
+	signHmac.Write([]byte(encryptedData))
+	verifySign := signHmac.Sum(nil)
+	signDecoded, err := hex.DecodeString(sign)
+	if err != nil {
+		return false
```
Comment (andrewslotin) [inline session/session.go]: I'm not really sure if an error should be silently dropped here. Did you consider returning it to the caller?

**Review by andrewslotin — CHANGES_REQUESTED** [review]
Hey, thanks again for your contribution! I made some comments on your changes, but overall the PR makes sense to me.

**Comment by gingerhot** [comment]
@andrewslotin Thank you for your review and comments in details, I really learned a lot! I've refactored the codes.

File: session/session.go [hunk 7]
```diff
@@ -18,20 +18,20 @@ func generateSecret(base, salt string) []byte {
...
 	if err != nil {
-		return false
+		return false, err
 	}
 	if hmac.Equal(verifySign, signDecoded) {
-		return true
+		return true, nil
 	}
-	return false
+	return false, errors.New("session: signature not verified")
```
Comment (andrewslotin) [inline session/session.go]: You might also consider using a guard-clause style, i.e.
```go
if !hmac.Equal(/* ... */) {
    return false, ErrInvalidSignature
}

return true, nil
```

File: session/session.go [hunk 8]
```diff
@@ -73,10 +73,11 @@ func DecryptSignedCookie(signedCookie, secretKeyBase, salt, signSalt string) (se
 
 	vectors := strings.SplitN(cookie, "--", 2)
 	if vectors[0] == "" || vectors[1] == "" {
-		return nil, errors.New("Invalid cookie!")
+		return nil, errors.New("session: invalid cookie")
 	}
-	if verified := verifySign(vectors[0], vectors[1], secretKeyBase, signSalt); !verified {
-		return nil, errors.New("Not verified signature!")
+	verified, err := verifySign(vectors[0], vectors[1], secretKeyBase, signSalt)
```
Comment (andrewslotin) [inline session/session.go]: Seems like this error was not handled.

File: session/session_test.go [hunk 9]
```diff
@@ -21,23 +21,28 @@ func TestVerifySign(t *testing.T) {
 	cookie, _ := url.QueryUnescape(signedCookie)
 	vectors := strings.SplitN(cookie, "--", 2)
 	// a right signature case
-	if verified := verifySign(vectors[0], vectors[1], secretKeyBase, signSalt); !verified {
-		t.Error("The function verifySign() test failure!")
+	verified, err := verifySign(vectors[0], vectors[1], secretKeyBase, signSalt)
```
Comment (andrewslotin) [inline session/session_test.go]: Same here, not sure if not handling errors is a good idea.

**Review by andrewslotin — CHANGES_REQUESTED** [review]

File: session/session.go (hunk 7 above)
Comment (gingerhot) [reply session/session.go]: Yes, guard-clause style is secure and right way.

File: session/session.go (hunk 8 above)
Comment (gingerhot) [reply session/session.go]: How about I change it as:

```go
if !verified || err != nil {
    return
}
```

because verified and err not nil will be consistent in the `verifySign` function. Or should I change it as follow more reasonable?

```go
if err != nil {
    return
}
if !verified {
    return nil, ErrInvalidSignature
}
```

Thanks~

File: session/session_test.go (hunk 9 above)
Comment (gingerhot) [reply session/session_test.go]: same as above.

File: session/session.go (hunk 8 above)
Comment (andrewslotin) [reply session/session.go]: I'd prefer the latter option to keep internal errors separate from user input errors (i.e. invalid signature)

Comment (gingerhot) [reply session/session.go]: But maybe we're not able to distinguish user input errors from internal errors.

If so the error duplicates:
```go
func verifySign(encryptedData, sign, base, signSalt string) (bool, error) {
   /* ... */
    if !hmac.Equal(verifySign, signDecoded) {
        return false, errors.New("session: signature not verified")
    }
    return true, nil
}
```

with the caller's checking:
```go
func DecryptSignedCookie(signedCookie, secretKeyBase, salt, signSalt string) (session []byte, err error) {
    /* ... */
    if err != nil {
        return
    }
    if !verified {
        return nil,  errors.New("session: signature not verified")
    }
    /* ... */
}
```

So I think maybe the former one is more clearer here.

Comment (andrewslotin) [reply session/session.go]: This is why I'd recommend to use [a sentinel value](https://en.wikipedia.org/wiki/Sentinel_value) in this case. You might consider `io.EOF` as an example:
```go
for {
    err := reader.Read()
    if err == io.EOF {
        break // not an actual error, but requires special handling
    } else if err != nil {
        log.Fatalf("error: %s", err)
    }

    // ...
}
```
So I'd propose to add `var ErrInvalidSignature = errors.New("session: signature not verified")` and return it instead of creating a new error.

Comment (gingerhot) [reply session/session.go]: Do you mean I'd better write it as below  to eliminate duplication by the `ErrInvalidSignature `?

```go
var ErrInvalidSignature = errors.New("session: signature not verified")
// ...

func verifySign(encryptedData, sign, base, signSalt string) (bool, error) {
   // ...
    if !hmac.Equal(verifySign, signDecoded) {
        return false, ErrInvalidSignature
    }
    return true, nil
}


func DecryptSignedCookie(/* ... */) (session []byte, err error) {
    // ...
    if err != nil {
        return nil , err
    }
    if !verified {
        return nil,  ErrInvalidSignature
    }
}
```

Thanks~

Comment (andrewslotin) [reply session/session.go]: This is what I had in mind. And the caller can easily assert if the returned error is actually `ErrInvalidSignature`.

Comment (gingerhot) [reply session/session.go]: Yes, this's much better than my original code. I'll push the new code later.

Thank you again for your efforts to show me the skill~

**Review by andrewslotin — APPROVED** [review]
Thanks again for your contribution
